p2p/connection.py

```diff
@@ -0,0 +1,324 @@
+import asyncio
+import collections
+import logging
+import operator
+from typing import (
+    AsyncIterator,
+    cast,
+    Any,
+    Dict,
+    Type,
+    Tuple,
+)
+
+from cached_property import cached_property
+
+import rlp
+
+from eth.tools.logging import ExtendedDebugLogger
+
+from p2p._utils import (
+    get_devp2p_cmd_id,
+)
+from p2p.constants import (
+    SNAPPY_PROTOCOL_VERSION,
+)
+from p2p.exceptions import (
+    DecryptionError,
+    HandshakeFailure,
+    MalformedMessage,
+    NoMatchingPeerCapabilities,
+    PeerConnectionLost,
+    TooManyPeersFailure,
+    UnknownProtocolCommand,
+)
+from p2p.kademlia import Node
+from p2p.p2p_proto import (
+    Command,
+    Disconnect,
+    DisconnectReason,
+    Hello,
+    P2PProtocol,
+    PayloadType,
+    Ping,
+)
+from p2p.protocol import (
+    empty_context,
+    BaseProtocolContext,
+    Protocol,
+)
+from p2p.transport import Transport
+
+
+TCapability = Tuple[str, int]
+
+
+class Connection:
+    logger = cast(ExtendedDebugLogger, logging.getLogger('p2p.connection.Connection'))
+
+    transport: Transport
+    _supported_protocols: Tuple[Type[Protocol], ...]
+    is_inbound: bool
+    protocol_context: BaseProtocolContext
+
+    # The following properties will be set uppon successful completion of the handshake
+    client_version_string: str
+    protocol: protocol.Protocol = None
+
+    # FIXME: Must be configurable.
+    listen_port = 30303
+
+    # Will be set upon disconnect
+    disconnect_reason: DisconnectReason = None
+
+    def __init__(self,
+                 transport: Transport,
+                 protocols: Tuple[Type[Protocol], ...],
+                 is_inbound: bool,
+                 context: BaseProtocolContext = None) -> None:
+        self.transport = transport
+        self._supported_protocols = protocols
+        self.is_inbound = is_inbound
+
+        if context is None:
+            context = BaseProtocolContext()
+        self.protocol_context = context
+
+        # Number of received messages for each message type.
+        self.received_msgs: Dict[Command, int] = collections.defaultdict(int)
+        self.ready = asyncio.Event()
+
+    #
+    # proxied Transport properties and methods
+    #
+    @cached_property
+    def remote(self) -> Node:
+        return self.transport.remote
+
+    @property
+    def is_closing(self) -> bool:
+        return self.transport.is_closing
+
+    def close(self) -> None:
+        self.transport.close()
+
+    def get_safe_client_version_string(self) -> str:
+        # limit number of chars to be displayed, and try to keep printable ones only
+        # MAGIC 256: arbitrary, "should be enough for everybody"
```

Comment:
This function mutated while being copy-pasted, and there's no longer a check for client-string-too-long. Here's how it used to look in handshake processing before this PR:

``` python
        # limit number of chars to be displayed, and try to keep printable ones only
        # MAGIC 256: arbitrary, "should be enough for everybody"
        original_version = msg['client_version_string']
        client_version_string = original_version[:256] + ('...' if original_version[256:] else '')
        if client_version_string.isprintable():
            self.client_version_string = client_version_string.strip()
        else:
            self.client_version_string = repr(client_version_string)
```

Now, `self.client_version_string` is set directly from the remote's handshake message (without sanitising).

At the very least, remove the comments, as they no longer represent truth.

My intent with this "length limiting" was to prevent [senseless crap](https://github.com/ethereum/trinity/pull/332#discussion_r262914939) being pushed from the remote.

IMO it's inappropriate to have this function: the string should already be sanitised when attached to an object instance.

Reply:
Good catch and yes, want to keep that sanitization.  Will get it re-added.